### PR TITLE
[rbrowser] Move RBrowserDataCleanup into header and LinkDef.h

### PR DESCRIPTION
**Pull request by hahnjo**
This fixes the test `roottest-root-core-execStatusBitsCheck` when building with `runtime_cxxmodules=OFF`, see #13058.

---

I don't know if there is an easier / better solution for this...

### CODE REVIEW

**Comment by linev** [comment]
Does it mean - any class derived from TObject should have ClassDef and dictionary?

Even if it helps here - there are many other libraries which do not create dictionaries.

**Comment by hahnjo** [comment]
> Does it mean - any class derived from TObject should have ClassDef and dictionary?

I'm not sure, we'll need @pcanal here. My understanding of the problem is that `RFileDialog` has a member of type `RBrowserData` (which has an entry in the `LinkDef.h`) which has a member of type `std::unique_ptr<RBrowserDataCleanup>`, where `RBrowserDataCleanup` is only forward declared.

**Comment by linev** [comment]
Member in RBrowserData.hxx defined as:
```
std::unique_ptr<RBrowserDataCleanup> fCleanupHandle;  ///<! cleanup handle for RecursiveRemove
```
It should be fully excluded from I/O. 

Only if dictionary has problem with forward declarations.

**Comment by linev** [comment]
> But my understanding is that this not IO complaining, but already TClass...

But again - it still should be possible to use such class definitions without moving specialization into header files. 
I expect, that ClassDef and LinkDef.h changes are not necessary - just moving `RBrowserDataCleanup` to header files.

**Comment by pcanal** [comment]
> I expect, that ClassDef and LinkDef.h changes are not necessary 

Technically since the class inherits from `TObject` it ought to have a `ClassDef` but it could/should be a `ClassDefInline`.

**Comment by pcanal** [comment]
I do not see anything inherently wrong with the changes BUT I really don't understand the failure and why moving the decl fixes them.  I.e. it feels that this is hiding an existing problem.  In particular, I see in https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/178429/testReport/projectroot.roottest.root/core/roottest_root_core_execStatusBitsCheck/:

```
+Error in <TProtoClass::FindDataMember>: data member with index 0 is not found in class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> >
+Error in <CreateRealData>: Cannot find data member # 0 of class 
```
Whereas I do not even see/find the use of `RBrowserDataCleanup` within a `std::tuple`  ...

I only see it used in a transient member:
```
origin/master:gui/browserv7/inc/ROOT/RBrowserData.hxx:   std::unique_ptr<RBrowserDataCleanup> fCleanupHandle;  ///<! cleanup handle for RecursiveRemove
```
which should be suppressing any warning about missing information (albeit not the one above that usually indicates an internal error)

**Comment by linev** [comment]
> I.e. it feels that this is hiding an existing problem.

Fully agree. One should try to analyze why problem appears at all.

**Comment by hahnjo** [comment]
> I expect, that ClassDef and LinkDef.h changes are not necessary - just moving RBrowserDataCleanup to header files.

I tested this before submitting, and it didn't fix the problem. But I'm by no means an expert in this area, entirely possible that I'm missing something!

> Fully agree. One should try to analyze why problem appears at all.

@pcanal could you help with this? I'm not very familiar with this part of ROOT...

**Comment by vepadulano** [comment]
My 2 cents:

1. It is extremly weird/ a bug that the webgui needs to depend on `TObject`.
2. `RBrowserDataCleanup` derives from `TObject`, so it *must* have a `ClassDef` as per https://root.cern/manual/io_custom_classes/#the-classdef-macro . I am not sure about what using  `ClassDefInline` as suggested by @pcanal would change
2a. Can `RBrowserDataCleanup` just avoid inheriting from `TObject` completely? 
3. `RBrowserData` has a dictionary and allows I/O, as it is defined in `LinkDef.h`, so it is only natural that its data members need a dictionary as well. As @hahnjo commented, the particular error coming from a `std::tuple` is due to the fact that the class uses ` std::unique_ptr<RBrowserDataCleanup>`.

Bottom line, to me everything looks very clear and I don't understand the comments regarding looking for a deeper reason behind the error. The only thing I don't understand is the dependency from `TObject` itself, but that's beyond the scope of this PR

**Comment by hahnjo** [comment]
My understanding is that `RBrowserDataCleanup` inherits from `TObject` to have it registered in `gROOT`, cf. commit cb8d8b73ea856aa7b895c38112bbefbcf0967cfa that introduced it.

A dictionary should, in fact, not be needed because the data members are transient. This is what needs to be investigated.

**Comment by linev** [comment]
> RBrowserDataCleanup derives from TObject, so it must have a ClassDef

Is it so? 
To my understanding, `ClassDef` required either for I/O and/or if class used with interpreter. 
Both is not required for `RBrowserDataCleanup`. 

Seems to be problem caused by dictionary generation for transient member `std::unique_ptr<RBrowserDataCleanup>`.
I can very easily solve the problem - replacing `std::unique_ptr` by plain pointer. But this will not solve problem with dictionary - any user can have similar class layout and will face problem again. Forcing the user to provide dictionary for the class which does not appear in headers? It is not very logical.

**Comment by vepadulano** [comment]
> Is it so?

Yes, as written in the manual. The manual could be wrong, but it was rewritten very recently so I expect this to be the most up-to-date information

**Comment by vepadulano** [comment]
FYI,

I tried the route of `ClassDefInline` with poor results. First off, the class now need a default constructor, which means I had to turn the reference data member to a pointer. The diff is here

```
--- a/gui/browserv7/src/RBrowserData.cxx
+++ b/gui/browserv7/src/RBrowserData.cxx
@@ -26,6 +26,8 @@
 #include "TBufferJSON.h"
 #include "TEnv.h"
 
+#include "TObject.h"
+
 #include <algorithm>
 #include <regex>
 
@@ -42,15 +44,15 @@ namespace Experimental {
 
 class RBrowserDataCleanup : public TObject {
 
-   RBrowserData &fData;
+   RBrowserData *fData{nullptr}; //!
 
 public:
-   RBrowserDataCleanup(RBrowserData &_data) : fData(_data) {}
+   RBrowserDataCleanup(RBrowserData *_data) : fData(_data) {}
+   RBrowserDataCleanup() {} // Needed by ClassDefInline
 
-   void RecursiveRemove(TObject *obj) override
-   {
-      fData.RemoveFromCache(obj);
-   }
+   void RecursiveRemove(TObject *obj) override { fData->RemoveFromCache(obj); }
+
+   ClassDefInline(RBrowserDataCleanup, 1);
 };
 }
```

And the test gives even more errors then before:

```
1548: -- BEGIN OUTDIFF OUTPUT --
1548: --- /home/vpadulan/Programs/rootproject/roottest/root/core/execStatusBitsCheck.ref	Wed Jun  9 11:08:31 2021
1548: +++ /home/vpadulan/Programs/rootproject/rootbuild/execstatusbitcheck-fix-v2/roottest/root/core/execStatusBitsCheck.log	Mon Jul 17 18:19:46 2023
1548: @@ -14,5 +14,29 @@
1548:  Bit  13 declared in TObject as kInvalidObject
1548:  Bit  14 declared in TStreamerElement as kWholeObject
1548:  Bit  21 declared in TStreamerElement as kWarned
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TClass::LoadClassInfo>: no interpreter information for class ROOT::Experimental::RBrowserDataCleanup is available even though it has a TClass initialization routine.
1548: +Error in <TProtoClass::FindDataMember>: data member with index 0 is not found in class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> >
1548: +Error in <CreateRealData>: Cannot find data member # 0 of class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> > for parent ROOT::Experimental::RFileDialog!
1548: +Error in <TProtoClass::FindDataMember>: data member with index 1 is not found in class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> >
1548: +Error in <CreateRealData>: Cannot find data member # 1 of class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> > for parent ROOT::Experimental::RFileDialog!
```

I guess the new errors come from the fact I am not generating a dictionary in `Linkdef.h`? Anyway, maybe this will give more food for thought to the conversation

**Comment by hahnjo** [comment]
Very likely you still want `ClassDefInline(RBrowserDataCleanup, 0)` to disable IO. I guess this is what requires the default constructor?

**Comment by linev** [comment]
>>    Is it so?

> Yes, as written in the manual. 

I still do not see reason why dictionary is necessary here. 
It is really question for @pcanal. 
Is there special handling of `std::unique_ptr<T>` for the types derived from `TObject`? 
In dictionary for transient members? 
And how one gets such information for the class which does not appears in header files?

As I mention - there is much simpler workaround with plain pointer.
But once I implement it - one immediately forgets about the problem.

**Comment by vepadulano** [comment]
> Very likely you still want ClassDefInline(RBrowserDataCleanup, 0) to disable IO. I guess this is what requires the default constructor?

Ah yes, I turned `0` to `1` while testing the other changes. Reverting back to `0` and keeping the rest of the diff as it is, I get again the usual error and no extra errors
```
1548: -- BEGIN OUTDIFF OUTPUT --
1548: --- /home/vpadulan/Programs/rootproject/roottest/root/core/execStatusBitsCheck.ref	Wed Jun  9 11:08:31 2021
1548: +++ /home/vpadulan/Programs/rootproject/rootbuild/execstatusbitcheck-fix-v2/roottest/root/core/execStatusBitsCheck.log	Mon Jul 17 18:29:44 2023
1548: @@ -14,5 +14,9 @@
1548:  Bit  13 declared in TObject as kInvalidObject
1548:  Bit  14 declared in TStreamerElement as kWholeObject
1548:  Bit  21 declared in TStreamerElement as kWarned
1548: +Error in <TProtoClass::FindDataMember>: data member with index 0 is not found in class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> >
1548: +Error in <CreateRealData>: Cannot find data member # 0 of class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> > for parent ROOT::Experimental::RFileDialog!
1548: +Error in <TProtoClass::FindDataMember>: data member with index 1 is not found in class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> >
1548: +Error in <CreateRealData>: Cannot find data member # 1 of class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> > for parent ROOT::Experimental::RFileDialog!
```

So no actual improvement brought by this change it seems

**Comment by pcanal** [comment]
> Is there special handling of std::unique_ptr<T> for the types derived from TObject?

No.  And internally the treatment of `unique_ptr<T>` and `T*` should be exactly the same.

**Comment by linev** [comment]
> And internally the treatment of unique_ptr<T> and T* should be exactly the same.

In worse case I can just use `void *`. But I 99.99% sure that plain pointer will not harm dictionary generation.

**Comment by pcanal** [comment]
> To my understanding, ClassDef required either for I/O and/or if class used with interpreter.

Not only.  If an object inherits from `TObject`, it gets the member functions `IsA` and `Class`.  If the derived class does have a `ClassDef` those will returns the wrong answer (i.e `this is object is a TObject` rather than `this object is a derived class`).  In practice this means that the behavior will be inconsistent and ***some things*** will fail silently (for example the Graphical interface will not be able to know which type object this is).

**Comment by pcanal** [comment]
>      Ah yes, I turned 0 to 1 while testing the other changes. Reverting back to 0 and keeping the rest of the diff as it is, I get again the usual error and no extra errors

I think those are a bug.  I am seeing it in a completely different setting.

**Comment by linev** [comment]
```
Error in <TProtoClass::FindDataMember>: data member with index 0 is not found in class tuple<ROOT::Experimental::RBrowserDataCleanup*,default_delete<ROOT::Experimental::RBrowserDataCleanup> >
```

Problem is this line. That is `data member with index 0` here?
`TClass::CreateRealData` tries interpret class member for which no information is provided.

**Comment by pcanal** [comment]
> I tried the route of ClassDefInline with poor results. 

Rigth ... When not creating a dictionary, the information about data member, etc. has to come from the interpreter at run-time .. i.e. it needs to loads the definition .... So indeed if the goal is to keep this class declared only in a source file, this is not the solution.

**Comment by pcanal** [comment]
> This unfortunately (sigh) fixes the error in the test

This indicates a bug in the code (that is suppose to make support for unique_ptr 'transparent').

**Comment by Axel-Naumann** [comment]
@pcanal I remember that you wanted TClass to InspectMembers unique_ptr and determine its data layout also for transient members, to make sure "the information is complete" or something. Here it seems that you would like to have this behavior changed?

**Comment by hahnjo** [comment]
The suggestion of having a header in `res/` doesn't work unfortunately because the payload code expects to `#include` everything. So we stick with the current change for now, to unblock the C++17 PR?

**Comment by pcanal** [comment]
>  So we stick with the current change for now, to unblock the C++17 PR?

Almost.  The class needs to be moved into the `ROOT::Experimental::Internal::` to document it as private to ROOT's implementation.

**Comment by pcanal** [comment]
> Here it seems that you would like to have this behavior changed?

It seems more complicated than that ... The underlying problem needs to be investigated.  Once this PR is merged, an issue will be open with words along the line of 'revert this commit, run this and see these errors messages`

**Comment by hahnjo** [comment]
> > So we stick with the current change for now, to unblock the C++17 PR?
> 
> Almost. The class needs to be moved into the `ROOT::Experimental::Internal::` to document it as private to ROOT's implementation.

Ok, so basically doing what @vepadulano had in https://github.com/root-project/root/pull/13257... Will do.

**Comment by hahnjo** [comment]
@phsft-bot build just on ubuntu2204/nortcxxmod with flags -DCMAKE_CXX_STANDARD=17

**Comment by hahnjo** [comment]
@phsft-bot build just on ROOT-ubuntu2204/nortcxxmod with flags -DCMAKE_CXX_STANDARD=17

**Comment by hahnjo** [comment]
@phsft-bot build just on ROOT-ubuntu2204/nortcxxmod with flags -DCMAKE_CXX_STANDARD=17

**Comment by hahnjo** [comment]
@phsft-bot build just on ROOT-ubuntu2204/nortcxxmod with flags -DCMAKE_CXX_STANDARD=17

**Review by vepadulano — APPROVED** [review]
Approving after discussion with @pcanal. The current faulty behaviour of TClass should be better investigated and is documented in an issue https://github.com/root-project/root/issues/13361 .The changes in the PR unblock the upgrade to C++17 and consequently to LLVM16

**Comment by Axel-Naumann** [comment]
Sorry - (too) late to the game. Why is the member not `std::unique_ptr<TObject> fCleanupHandle;` and we keep the definition of `RBrowserDataCleanup` in the source file?

**Comment by hahnjo** [comment]
> Sorry - (too) late to the game. Why is the member not `std::unique_ptr<TObject> fCleanupHandle;` and we keep the definition of `RBrowserDataCleanup` in the source file?

Yes, this also totally works... :expressionless: https://github.com/root-project/root/pull/13363

**Pull request merged**
